core/certificate.go

```diff
@@ -98,11 +98,11 @@ func (s *CertificatesClient) doRequest(
 ) (*certificatesResponseBody, *katapult.Response, error) {
 	u = s.basePath.ResolveReference(u)
 	respBody := &certificatesResponseBody{}
-	resp := katapult.NewResponse(nil)
 
-	req, err := s.client.NewRequestWithContext(ctx, method, u, body)
-	if err == nil {
-		resp, err = s.client.Do(req, respBody)
+	req := katapult.NewRequest(method, u, body)
+	resp, err := s.client.Do(ctx, req, respBody)
+	if resp == nil {
+		resp = katapult.NewResponse(nil)
 	}
```

Comment:
Instead of potentially returning `nil`, can't we return `NewResponse(nil)` so we don't have to check and set the response every time we use the client? Maybe there is some go idiom I'm not aware of at play... 😅

Reply:
I've actually forgotten the reason for this here, so if @jimeh could remind that would be good. I know we are hopefully eventually going to get to a point in the future where ``nil`` is returned when an error occurs, so perhaps this is a bit of a midway point.

Reply:
I actually used to ensure that `Do` always returned a non-nil response, but it meant these checks were never executed. And technically without them there's a risk of a panic if `Do` for any reason did return a nil response.

This comes down to the Pagination data that we get from the body structs and attach to the response object.

The response object itself is currently needed to return error details, and pagination info. With #64 we'll get error details as regular Go errors, removing one reason for the response object to exist. And pagination info we should start returning as some form of meta/pagination return value on list endpoints. We should probably have a discussion about the best way to handle pagination info.

But yes, in the short term, I thought it was safest to have this check, and to let `Do` return a nil response, so there's explicit behavior that's accounted for, and not just a convention of never returning nil while it's a perfectly valid return value as far as Go is concerned.